---
title: "Analyse d'expression différentielle mutli-facteurs : Fer, Azote et CO2"
subtitle: "Etude du facteur fer"
author: "Océane Cassan"
date: "10/25/2019"
output: 
  rmdformats::material:
    fig_width: 12
    highlight: kate
    includes:
        after_body: footer.html
---

On cherche à étudier l'effet de trois facteurs sur le transcriptome des racines d'Arabidopsis thaliana.
Le CO2, au cours des études préliminaires, s'est montré peu influent en conditions contrôles de fer et de nitrates, et accentué en cas de stress nutritionnel.
Nous reprennons ces résultats avec des fonctions génériques et propres pour en faire le résumé et de jolis graphes.


```{r knitr_init, echo=FALSE, cache=FALSE}
library(knitr, warn.conflicts = F, quietly = T)
library(rmdformats, warn.conflicts = F, quietly = T)

## Global options
options(max.print="75")
opts_chunk$set(cache=FALSE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE,
               fig.align = "center")
opts_knit$set(width=75)
#setwd("~/Documents/CombinatoireFeNCO2/Combinatoire/")
library(gridExtra)
suppressMessages(library(gridExtra, warn.conflicts = F, quietly = T))
suppressMessages(library(stringr, warn.conflicts = F, quietly = T))
suppressMessages(library(ggpubr, warn.conflicts = F, quietly = T))
suppressMessages(library(ggplot2, warn.conflicts = F, quietly = T))
suppressMessages(library(clusterProfiler, warn.conflicts = F, quietly = T))
suppressMessages(library(enrichplot, warn.conflicts = F, quietly = T))

library(reshape2)

#self written functions for differential expression
source("./DEFunctions.R")
Path = "~/Documents/CombinatoireFeNCO2/DEG_Combinatoire/MicroTom/FactorIron/"

```


# Import des données : matrice d'expression

On a, pour chaque gène et chaque condition, son niveau d'expression en sortie de quantification.
On labelle les conditions avec le code suivant : lettre majuscule pour le niveau fort, minuscule pour le niveau faible. Le réplicat est donné après l'underscore.

```{r data}
#quantification file
data <- read.csv("quantifFiles/QuantifGenes.csv", h = T, sep = ',')
rownames(data) <- data$Gene
genes = which(!(grepl("__", rownames(data))))
not_quant = data[which((grepl("__", rownames(data)))),]
data = data[genes,grepl('R', colnames(data))]

getLabel("R6")
getExactCondition("R4")
getCondition("R4")

keep <- rowSums(data) >= 10
data <- data[keep,]
group <- sapply(colnames(data), getLabel, with.rep = F)
colnames(data) <- sapply(colnames(data), getLabel)
head(data)
dim(data)

specie = "At"
```



# Effet fer en conditions contrôle

On définit les conditions contrôle comme suit : CO2 ambiant et fort fer.

```{r contr}
g = list()
method = "edger"
labels <- c("cNF", "cNf")
genes1 <- dualDE(data, labels, pval = 0.01, method=method)
g[[paste(labels[1], labels[2])]] = genes1

a <- OntologyProfile(genes1$gene_id, specie = specie)
b <- writeGenes(export = Path, labels = labels, ont = a, DEresult = genes1)
```

# Fort CO2

```{r lowN}
labels <- c("CNF", "CNf")
genes2 <- dualDE(data, labels, pval = 0.01)
g[[paste(labels[1], labels[2])]] = genes2

a <- OntologyProfile(genes2$gene_id, specie = specie)
b <- writeGenes(export = Path, labels = labels, ont = a, DEresult = genes2)
```

# Low nitrate

```{r lowF}
labels <- c("cnF", "cnf")
genes3 <- dualDE(data, labels, pval = 0.01)
g[[paste(labels[1], labels[2])]] = genes3

a <- OntologyProfile(genes3$gene_id, specie = specie)
b <- writeGenes(export = Path, labels = labels, ont = a, DEresult = genes3)
```

# Low nitrate and high CO2

```{r lowBoth}
labels <- c("CnF", "Cnf")
genes4 <- dualDE(data, labels, pval = 0.01)
g[[paste(labels[1], labels[2])]] = genes4

a <- OntologyProfile(genes4$gene_id, specie = specie)
b <- writeGenes(export = Path, labels = labels, ont = a, DEresult = genes4)
```


# Venn diagram

On visualise les gènes différentiellement exprimés en commun entre les différents niveaux des autres facteurs.

```{r venn}

library(ggVennDiagram)
library(VennDiagram)

gene_list <- list()
for(comp in names(g)){
  print( g[[comp]])
  gene_list[[comp]] <- g[[comp]]$gene_id
}
ggVennDiagram(gene_list)


partitions <- get.venn.partitions(gene_list, keep.elements = T)
partitions$shared <- rowSums(partitions[1:4])
partitions <- partitions[order(-partitions$shared),]

# common_genes <- unlist(partitions[1, "..values.."])
# results <- getBM( filters = "ensembl_gene_id", attributes = c("ensembl_gene_id", "description", "external_gene_name", "entrezgene_id"),
#                     values = common_genes, mart = mart)
# results <- results[!rownames(results) %in% which(duplicated(results$ensembl_gene_id)), ]
# kable(results)
# 
sharedBy3 <- unique(unlist(subset(partitions, partitions$shared==3)$..values..))
a <- OntologyProfile(sharedBy3, specie = specie)
save(sharedBy3, file = paste0("GenesIron_", specie, ".RData"))
```

# Summary graph

```{r barplot}
d <- data.frame(matrix(ncol = 4, nrow=2))
colnames(d) <- names(g)
row.names(d) <- c("up", "down")
for(comp in names(g)){
  d["up", comp] <- sum(g[[comp]]$upreg == 1)
  d["down", comp] <- sum(g[[comp]]$upreg == 0)
}
res <- melt(d)
res$reg = rep(c("up", "down"), 4)

genesIronAt <- res
save(genesIronAt, file = paste0("genesIron", specie, ".RData"))

ggplot(res, aes(fill=reg, y=value, x=variable)) + 
    geom_bar(position="stack", stat="identity", alpha=0.5, color = "black") +
    ggtitle("Iron effet on gene regulation") +
    xlab("") + ylab("Number of differentially expressed genes") + coord_flip()
```